### PR TITLE
Environment + Inference Overhaul

### PR DESCRIPTION
**Pull request by jakegrigsby**
This is a large update meant to make AMAGO faster, more focused on accelerated envs, and more usable for people outside our lab. It has a ton of breaking changes... which I'd like to stop doing going forward. 

The main additions are:
➕ Policy inference is faster, especially for the Transformer k/v cache
➕ Environment parallelization is more efficient
➕ Added experimental support for jax-style vectorized envs and an example script for [gymnax](https://github.com/RobertTLange/gymnax).
➕ Added a vectorized gym version of [XLand Minigrid](https://github.com/dunnolab/xland-minigrid). There are limits to how fast this can be in a pytorch agent, but this is a cool environment we'd like to support going forward.
➕ New README and Tutorial
➕ Public `wandb` examples (still in progress)
➕ Some mild corrections to the public version of the `MultiTaskAgent` for release alongside the [NeurIPS 2024 paper about this](https://openreview.net/pdf?id=OSHaRf4TVU).
➕ K/V cached inference without using FlashAttention. Fall back to vanilla pytorch attention when Flash isn't installed.
➕ Experimental support for [Flex Attention](https://pytorch.org/blog/flexattention/).
➕ Lots of other cleanups
➕ More torch compilation by default.
➕ Custom Transformer changes discussed in the first paper are easier to configure.

There are some losses:
➖ **Hindsight Relabeling**. This is temporary. The whole env setup was built to do this but had to be rewritten. We've left the door open to bringing this back by putting the goals in the observation dict. 
➖ Crafter, "Delivery" and "MazeRunner" envs from the ICLR 2024 paper that relied on hindsight relabeling. Again, we plan on brining this back.
➖ "Slow inference mode". We used to be able to cut the the gap between training/testing by storing the entire context sequence during inference. This was mainly for testing new sequence models, hasn't been used in a long time, and really slowed things down.

Despite the huge diff there are only a few things that should impact reproducibility of old (non-relabeling) results:
❓  Changed the initialization of SigmaReparam layers to more closely match the official version. Ours was written before that code was released, and led to large values on init. We were getting bailed out by FlashAttention and other stability tricks.
❓  Removed time information from the policy input. We could still put this in the observation. This removes the need to set a `horizon` in every experiment, which was confusing. Without time info, we now do infinite bootstrapping by only setting done=True on `terminated = True` (not `truncated = True`). 

We've done a lot of testing to make sure these edits don't break key results and will keep these replication runs going as we build a public `wandb` of reference logs.

### CODE REVIEW

**Comment by jakegrigsby** [comment]
**CartPole-v0** (2 x 256d tformer, l=128, batch=24, 128 async actors): `main`: 3900 FPS, 33 updates/sec; `refactor-gc` 5650 FPS, 38 updates/sec
**LunarLanderV2** (6 x 256d tformer, l=128 batch=24, 64 async actors): `main`: 2200 FPS, 19 updates/sec; `refactor-gc`: 2340 FPS, 21 updates/sec 
**HalfCheetahV4** (6 x 256d tformer, l=128 batch=24, 64 async actors): `main`: 1480 FPS, 15 updates/sec; `refactor-gc` 2100 FPS, 17 updates/sec
**10-Game Atari** (6 x 256d tformer w/ impala, l=128 batch=24, 60 async actors): `main` 860 FPS, 3.75 updates/sec; `refactor-gc` 1200 FPS, 4.2 updates/sec
**BabyAI** (3 x 256d tformer, l=600 batch=24, 60 async actors): `main` 1520 FPS, 22 updates/sec, `refactor-gc` 2250 FPS, 20 updates/sec 
**MetaWorld ML1 (reach-v2)** (3 x 384d tfrormer, l=600 batch=24, 60 async actors): `main` 1270 FPS, 5 updates/sec; `refactor-gc` 2030 FPS, 8.1 updates/sec

File: amago/envs/builtin/xland_minigrid.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        self.ruleset_benchmark = ruleset_benchmark
+        self.parallel_envs = parallel_envs
+        self.jax_device = (
+            jax.devices()[jax_device]
+            if jax_device is not None
+            else jax.devices("cpu")[0]
+        )
+        print(f"Using JAX device {self.jax_device} for XLandMiniGridEnv")
+
+        # this always uses cuda memory for some reason?
```
Comment (Howuhh) [inline amago/envs/builtin/xland_minigrid.py:71]: Yeah, this is just the default jax behaviour, i.e. putting all jax arrays to accelerator if available. You can put it to cpu manually. For example with `with jax.default_device(cpu_device):`.

File: amago/envs/builtin/xland_minigrid.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+    def get_obs(self):
+        obs = self.x_timestep.observation
+        done = self.x_timestep.last()
+        direction_done = np.concatenate(
+            (obs["direction"], done[:, np.newaxis]), axis=-1
+        )
+        # it seems more in the spirit of AdA's XLand to give the "goal" as an observation
+        # but let the agent meta-learn the "rules"... though it's not clear if that
+        # makes xland minigrid easier than it was meant to be...
```
Comment (Howuhh) [inline amago/envs/builtin/xland_minigrid.py:161]: From my experience with simple RL^2 baseline, explicit conditioning on true rules and goals makes it a lot easier. While AdA indeed revealed them, they used random masking during training + completely masked it out on evaluation (which is I think right thing to do).

File: amago/envs/builtin/xland_minigrid.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        if done.any():
+            self.reset_if_done(done)
+
+        next_obs = self.get_obs()
+        reward = np.array(reward, dtype=np.float32)
+        done = np.array(done, dtype=np.bool_)
+        return next_obs, reward, done, done, info
+
+    def render(self, env_idx: int, *args, **kwargs):
+        # this does not work with numpy > 2.0
```
Comment (Howuhh) [inline amago/envs/builtin/xland_minigrid.py:248]: hopefully fixed it...

**Comment by jakegrigsby** [comment]
Thanks @Howuhh! Not sure how you found this but was going to email you about it at some point if I could make some progress on the xland minigrid tasks beyond trivial-1m. 

The device placement / jax gpu memory situation should hopefully be fixed.

If i'm understanding right, AdA is randomly masking the "production rules" but not the "goal" (based on Table C1 in the appendix). Is it fair to interpret goals vs. rules in xland minigrid the same way? So using the `RulesAndGoalsObservationWrapper` but only passing the goal encoding to the observation (dropping the rule encoding) would be similar info as AdA gets when the production rules are fully masked?

I'm curious if you have some guesses on the memory lengths / model sizes / architectures that would be needed to make progress on the harder task benchmarks. It seems like the max episode length for the default 25-shot x 4 room x 13 grid setup is longer than most of the AdA experiments. AdA's models needs 1800-4200 timesteps of context, how do you think that compares to the minigrid version? Would these tasks take 25 trials to explore? That's much longer than I'd expect an RNN to stand a chance. Above 4k might be too expensive to try flat transformers... but I have mamba and am thinking about adding TXL or sparse attention if necessary.

**Comment by Howuhh** [comment]
Hi @jakegrigsby. Actually, I'm a big fan of amago and have been following it since the beginning! The amount of practical useful detail in the appendix of the paper is enormous...

Yes, I double-checked, and Ada really only masks rules, not goals, so your suggestion is totally valid. For some reason I just didn't notice the difference when I read it in previous times.

As for the number of trials / length of episodes, I think it can be guessed that all current numbers are at most educated guesses, but not final. Thus, I would actually be extremely grateful for feedback with how this works for a much more advanced method like Amago. From my very recent experience with simple RL^2, majority of problems in trivial can be "solved" with 2-3 trials, in small & medium with 3-5 trials. Of course it's all about mean values, unfortunately generalization (median/ 20th percentile) is always almost around zero.

As for hard benchmark, it's hard to say. It is quite diverse. There are both simple tasks and extremely, extremely difficult ones. Mostly their complexity is not something intellectual, of course, but rather very high memory requirements, because there are a lot of objects and rules that need to be searched, and for this it is important to remember that the agent has already tried. At the moment, I am actually more inclined that such complex tasks should be used rather to test generalization. Basically, to train on simpler ones, and on eval (so as not to slow down learning too much with large horizons), give a much larger budget for such complex tasks and check if the agent can generalize on them. Because essentially they are not much different from the medium problems, there are just more of the same rules + tree depth. Ideally, if the agent has really figured out the pattern that these tasks are built on, then for hard it just needs more time, but not some qualitative superiority.

**Comment by jakegrigsby** [comment]
@Howuhh hmm I see, thanks! I've been running some basic tests on trivial-1m and small-1m on the 1 room 9 grid size with mamba and 2048 timesteps of context. This seems to be working fine. I'll let you know if I can get some signs of life on 4x13 small-1m/medium-1m soon. I did add logging for the bottom 20% of returns -- I see why that would be important for the procedurally generated tasks. If mamba length 2048 is not enough, the next step would be adding sparse attention to imitate the 4-timestep downsampling used in AdA (or possibly something fancier than that). 

The obvious problem is that this repo is designed for generic gym envs and will never get the throughput of the end-to-end jax ecosystem. After all this refactoring I can comfortably do 50-100k FPS, which is a ton by my standards, but far slower than you're getting on your RL^2 baseline. 

Do you think there is a path towards a simpler version of AdA's auto-curriculum for sampling the next ruleset? Ideally without having to evaluate the current agent vs. no-op actions. Could we create a rough sorting of task difficulty instead of uniform sampling?

**Comment by Howuhh** [comment]
@jakegrigsby 50-100k FPS is a lot! My baselines are extremely tiny compared to amago. The bigger the model, there more it becomes a bottleneck rather than an environment, so for models of comparable size even end2end jitted training will give similar FPS from my experience. 

As for the rulesets curriculum, unfortunately I haven't had time to look into it myself yet. Overall, I think some form of it is crucial for a success and generalization.

**Comment by jakegrigsby** [comment]
Thanks @Howuhh, I'll try to add some kind curriculum if progress on xland minigrid stalls out.
